### PR TITLE
jwt_authn: Add subject constraints for JwtProviders

### PR DESCRIPTION
**Pull request by mattjo**
Commit Message: Add subject constraints for JwtProviders.
Additional Description: Adds new `subjects` config field to restrict subjects accepted from a `JwtProvider` partially implementing #31455
Risk Level: Low
Testing: Unit testing
Docs Changes: Added `subjects` description inline in proto.
Release Notes: Attached
Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md): Feature is opt in, without specifying the config, there's no behavior change.

### CODE REVIEW

**Comment by mattjo** [comment]
Hi @lizan and @htuch! Just wanted to check that this PR is on the review queue.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:120]: This makes sense to me from an API perspective. @lizan ?

**Comment by KBaichoo** [comment]
Friendly ping @lizan PTAL

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto
Comment (lizan) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:120]: Yes this makes sense.

File: source/extensions/filters/http/jwt_authn/jwks_cache.h
Comment (lizan) [inline source/extensions/filters/http/jwt_authn/jwks_cache.h:60]:
```suggestion
    virtual bool isSubjectAllowed(absl::string_view sub) const PURE;
```

nit: no top-level const in declaration.

**Comment by mattjo** [comment]
Thanks for the comment! Accepted change and resolved merge conflicts in current.yaml

**Comment by mattjo** [comment]
@htuch and @lizan PTAL, I think this is ready for approval

**Review by lizan — APPROVED** [review]

**Comment by lizan** [comment]
can you resolve the conflicts?

**Comment by mattjo** [comment]
> can you resolve the conflicts?

Done!

**Comment by mattjo** [comment]
@lizan Resolved current.yaml again, looks to be pretty high change rate so it's hard to keep clean.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by mattjo** [comment]
@lizan looks like you have to approve again after the merge, PTAL

**Review by lizan — APPROVED** [review]

**Pull request merged**
